v3_usm.go

```diff
@@ -622,28 +622,89 @@ func (sp *UsmSecurityParameters) discoveryRequired() *SnmpPacket {
 }
 
 func (sp *UsmSecurityParameters) calcPacketDigest(packet []byte) []byte {
-	var mac hash.Hash
+	return calcPacketDigest(packet, sp)
+}
 
-	switch sp.AuthenticationProtocol {
-	default:
-		mac = hmac.New(crypto.MD5.New, sp.SecretKey)
-	case SHA:
-		mac = hmac.New(crypto.SHA1.New, sp.SecretKey)
-	case SHA224:
-		mac = hmac.New(crypto.SHA224.New, sp.SecretKey)
-	case SHA256:
-		mac = hmac.New(crypto.SHA256.New, sp.SecretKey)
-	case SHA384:
-		mac = hmac.New(crypto.SHA384.New, sp.SecretKey)
-	case SHA512:
-		mac = hmac.New(crypto.SHA512.New, sp.SecretKey)
-	}
+// calcPacketDigest calculate authenticate digest for incoming messages (TRAP or
+// INFORM).
+// Support MD5, SHA1, SHA224, SHA256, SHA384, SHA512 protocols
+func calcPacketDigest(packetBytes []byte, secParams *UsmSecurityParameters) []byte {
+	var digest []byte
+
+	switch secParams.AuthenticationProtocol {
+	case MD5, SHA:
+		digest = digestRFC3414(
+			secParams.AuthenticationProtocol,
+			packetBytes,
+			secParams.SecretKey)
+	case SHA224, SHA256, SHA384, SHA512:
+		digest = digestRFC7860(
+			secParams.AuthenticationProtocol,
+			packetBytes,
+			secParams.SecretKey)
+	}
+
+	return digest
+}
 
+// digestRFC7860 calculate digest for incoming messages using HMAC-SHA2 protcols
+// according to RFC7860 4.2.2
+func digestRFC7860(h SnmpV3AuthProtocol, packet []byte, authKey []byte) []byte {
```

Comment:
Can you update the function to additionally return the possible ``mac.Write(packet)`` error:
```go
func digestRFC7860(h SnmpV3AuthProtocol, packet []byte, authKey []byte) ([]byte, error) {
	mac := hmac.New(h.HashType().New, authKey)
	_, err := mac.Write(packet)
	msgDigest := mac.Sum(nil)
	return msgDigest, err
}
```